### PR TITLE
Do not merge (wait for Julia 0.7): move type recipe definition to recipesbase

### PR DESCRIPTION
**Pull request by piever**
This is a follow up to the discussion in [#138](https://github.com/JuliaPlots/StatPlots.jl/issues/138). The first step is moving the type definition here, then when this PR is merged and there's a release, StatPlots can be updated to overload these types to define the recipes.

The advantage is that if a package want to use say a `groupedbar` in a recipe, they can simply depend on RecipesBase and return a `GroupedBar` object. The user will need to load StatPlots to plot it but the package does not need to take the dependency (if the user does not load StatPlots there will be a helpful error message).

With a similar reasoning I've added here `group_as_matrix` which recipe types need to overload to achieve a specific type of grouping behavior. In this way such recipe types can be defined without depending on Plots. Also in this case, the change does not actually do anything, but Plots in the future will extend `RecipesBase.group_as_matrix` rather than defining its own.

### CODE REVIEW

**Comment by mkborregaard** [comment]
We've let previous updates of RecipesBase coincide with julia releases to avoid ever having dependency clashes. Do you think that's reasonable here too?

**Comment by piever** [comment]
That's a good point. This feature is not crucial as people can do tricks with Require in the meantime and hopefully Julia 0.7 is not too far away, so maybe we can wait till Julia 0.7.

We just need to remember to add type recipes in this branch as they get added to StatPlots.

File: src/typedefs.jl [hunk 1]
```diff
@@ -0,0 +1,21 @@
+@userplot CornerPlot
```
Comment (mkborregaard) [inline src/typedefs.jl]: This is a nice and easy implementation of keeping these in RecipesBase. BUT it drastically increases the number of exported names from the package (It's normally only `apply_recipe`, `userplot`, `shorthands` and `series`), thus increasing the risk of clashes. What about just defining the type, then having `@userplot` look in RecipesBase for the type and register the rest of the functionality?

**Comment by piever** [comment]
It's true, we shouldn't export too much from this one, but doing `@userplot` from StatPlots wouldn't work as `@userplot` redefines the type. We can either define the type and shorthands in RecipesBase and sidestep `@userplot` entirely or, given that we're changing things, add an option to `@userplot` to not export the shorthands.

**Comment by mkborregaard** [comment]
Yes my suggestion was to change the behaviour of `userplot` by adding this possibility, either as an option or an automatic check.

**Comment by mkborregaard** [comment]
@piever is it time to revisit this?

**Comment by piever** [comment]
Yes! Will look at this in a few days.

**Comment by daschw** [comment]
Nice, thanks @piever!

**Pull request merged**